### PR TITLE
Add code samples for GAE Python bundled APIs.

### LINKED ISSUE #7890: Port code samples of GAE Python bundled APIs from GCS
## In which file did you encounter the issue?
N/A. I'm about to add a set of code samples.

### Did you change the file? If so, how?

Yes. `gsutil cp gs://python3-wormhole-api-samples/python3-wormhole-samples.zip .`

## Describe the issue
GAE Python bundled APIs went GA in April. The code samples for these APIs are still in a GCS bucket. 
https://storage.googleapis.com/python3-wormhole-api-samples/python3-wormhole-samples.zip

This issue is to publish these samples to 
[python-docs-samples](https://github.com/GoogleCloudPlatform/python-docs-samples)

### PR DESCRIPTION
**Pull request by jinglundong**
## Description

Fixes #7890 7890

Note: It's a good idea to open an issue first for discussion.

## Checklist
- [x] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [x] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [ ] **Tests** pass:   `nox -s py-3.6` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [x] Please **merge** this PR for me once it is approved.
- [x] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample

### CODE REVIEW

File: .github/CODEOWNERS
Comment (dandhlee) [inline .github/CODEOWNERS:21]: Is there someone from your team or your GitHub team that we can add as the owner?

File: appengine/standard_python3/bundled-services/blobstore/django/README.md
Comment (dandhlee) [inline appengine/standard_python3/bundled-services/blobstore/django/README.md:1]: Should this say anything specific for Django?

File: appengine/standard_python3/bundled-services/blobstore/flask/README.md
Comment (dandhlee) [inline appengine/standard_python3/bundled-services/blobstore/flask/README.md:1]: Ditto

File: appengine/standard_python3/bundled-services/blobstore/wsgi/README.md
Comment (dandhlee) [inline appengine/standard_python3/bundled-services/blobstore/wsgi/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/deferred/django/README.md
Comment (dandhlee) [inline appengine/standard_python3/bundled-services/deferred/django/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/deferred/flask/README.md
Comment (dandhlee) [inline appengine/standard_python3/bundled-services/deferred/flask/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/deferred/wsgi/README.md
Comment (dandhlee) [inline appengine/standard_python3/bundled-services/deferred/wsgi/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/mail/django/README.md
Comment (dandhlee) [inline appengine/standard_python3/bundled-services/mail/django/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/mail/flask/README.md
Comment (dandhlee) [inline appengine/standard_python3/bundled-services/mail/flask/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/mail/wsgi/README.md
Comment (dandhlee) [inline appengine/standard_python3/bundled-services/mail/wsgi/README.md:1]: Ditto.

**Review by dandhlee — CHANGES_REQUESTED** [review]
Please address lint issues.

As well, will there be any guides / instructions provided on how to run these samples for the users?

File: .github/CODEOWNERS
Comment (ace-n) [reply .github/CODEOWNERS:21]: @jinglundong LMK if you'd like help creating a GitHub team. 🙂

Comment (ace-n) [reply .github/CODEOWNERS:21]: **Update:** we can use the `@GoogleCloudPlatform/runtimes` team here.

@jinglundong cool if we add that team here?

File: appengine/standard_python3/bundled-services/blobstore/django/README.md
Comment (jinglundong) [reply appengine/standard_python3/bundled-services/blobstore/django/README.md:1]: Update the links to the corresponding page in our public doc. I don't want to repeat the doc here, since they will easily go out of sync.

File: appengine/standard_python3/bundled-services/blobstore/flask/README.md
Comment (jinglundong) [reply appengine/standard_python3/bundled-services/blobstore/flask/README.md:1]: Ditto

File: appengine/standard_python3/bundled-services/blobstore/wsgi/README.md
Comment (jinglundong) [reply appengine/standard_python3/bundled-services/blobstore/wsgi/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/deferred/django/README.md
Comment (jinglundong) [reply appengine/standard_python3/bundled-services/deferred/django/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/deferred/flask/README.md
Comment (jinglundong) [reply appengine/standard_python3/bundled-services/deferred/flask/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/deferred/wsgi/README.md
Comment (jinglundong) [reply appengine/standard_python3/bundled-services/deferred/wsgi/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/mail/django/README.md
Comment (jinglundong) [reply appengine/standard_python3/bundled-services/mail/django/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/mail/flask/README.md
Comment (jinglundong) [reply appengine/standard_python3/bundled-services/mail/flask/README.md:1]: Ditto.

File: appengine/standard_python3/bundled-services/mail/wsgi/README.md
Comment (jinglundong) [reply appengine/standard_python3/bundled-services/mail/wsgi/README.md:1]: Ditto.

**Review by jinglundong — COMMENTED** [review]
Updated all the README.md files to include the most relevant docs. Fixed all the formatting issues.

**Review by dandhlee — COMMENTED** [review]
Are there any tests to ensure that the samples continue to work properly in the future?

**Comment by jinglundong** [comment]
> Are there any tests to ensure that the samples continue to work properly in the future?

The short answer is no. I'm afraid that we need to track the end-to-end tests separately. 

I assume you referred to end-to-end tests, because unit tests aren't going to provide too much value here. Speaking of end-to-end test, do you know how /appengine and other components run their end-to-end tests? I don't see actions that run tests periodically (https://github.com/GoogleCloudPlatform/python-docs-samples/actions). 

https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/run/helloworld/e2e_test.py might be a good example (because it shows how to run a test using gcloud), but I don't know if we run them as part of the CI/CD.

**Comment by ace-n** [comment]
@jinglundong in our experience, deploying samples to a platform itself can often introduce significant additional flakiness. (This is particularly true with App Engine Flexible; perhaps Standard is nicer about this.)

Would testing these require deployment to App Engine (Standard), or can a developer run them with Django/Flask/WSGI on their local machine? If the latter, I wonder if a local "integration test" would be a good "middle ground" between unit tests and full-on E2E tests.

(*P.S:* I'm inclined to agree with @dandhlee that this PR itself should also include tests - even if they're only unit tests [and more comprehensive ones are a follow-up item]. That said, I'll defer to @iennae if she thinks otherwise.)

**Comment by jinglundong** [comment]
> deploying samples to a platform itself can often introduce significant additional flakiness.

I fully agree. Integration and end-to-end tests are more flaky than unit tests. 

> Would testing these require deployment to App Engine (Standard), or can a developer run them with Django/Flask/WSGI on their local machine?

These apps require a real deployment to App Engine Standard. We don't have devappserver2 support at this point I believe. Meanwhile, unit tests shouldn't be done at this layer. We have unit tests and end-to-end tests to guard SDK changes. Like here: http://shortn/_yZpYJzJc88 (I made it an internal link intentionally). 

> The goal is to ensure that the sample code can continue to run in the future, and if something does break we notice it through tests and dependency updates rather than users going through the bad experience and attempting to troubleshoot themselves.

I like the idea of having tests for code samples. There are at least two ways to break code samples. 1) when we change the code samples. 2) when the SDK and server-side behavior changes. For 1), before we have automated tests, we need to rely on manual tests. For 2), we have SDK tests and end-to-end tests before we make changes to the SDK or the service. So, I think 2) is out of the scope here. It's a production issue, and we will react fast. 

> if you can mimic what the AppEngine is doing with Blobstore... without the full deployment to the AppEngine, that should suffice the middle ground.

My intuition is that this type of tests will keep passing when the underlining system or code sample is broken. The end-to-end tests provides the highest confidence, but they are also hard to write and hard to de-flake. In the end, here's a trade-off between velocity and test coverage. Our team has very limited capacity comparing to the amount of products we own. If end-to-end tests are required, I think this PR will be delayed for a few weeks in reality.

**Comment by ace-n** [comment]
@jinglundong I see. @iennae do you have any thoughts WRT the timelines-vs-test-coverage tradeoff?

(_My personal opinion:_ I'd vote in favor of having better test coverage [or at minimum a follow-up issue], even if that takes longer.)

**Comment by iennae** [comment]
I think we get these included and then work towards figuring out the testing coverage. Right now these samples are being included in docs from a bucket. The first incremental improvement is to have them in GitHub. Second incremental improvement would be adding tests.

**Comment by ace-n** [comment]
@iennae SGTM. @dandhlee does that work for you?

**Comment by dandhlee** [comment]
Do these samples need to be published immediately? If not, then I hesitate approving for adding in samples without tests, just like adding code into production without test coverage... unless absolutely necessary. I'm more than faithful that, if something were to break between this being submitted then GA and tests being available, one of you will be more than capable of making the necessary updates. That could go unnoticed for a bit of time, is my concern. 

To sum this up (sorry for long blocky texts),

* If this needs to be published immediately, please file an issue to track for test coverage to be added in the near future.
* If this does not need to be published immediately and we can have _some_ test coverage, please file an issue to track for end-to-end test coverage, update the PR with some tests and I'll stamp and merge.
* If this does not need to be published immediately but having test coverage that is not end-to-end is expensive and difficult and does not provide meaningful checks, please file an issue to track, and if there's a consensus that time is better spent on maintaining these samples + coming up with end-to-end tests, I'll stamp and merge.

**Comment by ace-n** [comment]
> If this does not need to be published immediately but having test coverage that is not end-to-end is expensive and difficult and does not provide meaningful checks, please file an issue to track, and if there's a consensus that time is better spent on maintaining these samples + coming up with end-to-end tests, I'll stamp and merge.

I think this best reflects my personal opinion.

That, and this PR feels pretty large as-is. (Again - personally) I'd rather **a)** land this PR as an "intermediate milestone" and **b)** consider adding tests as a `P0 follow-up issue` rather than trying to do everything in one PR.

**Comment by jinglundong** [comment]
I filed a tracking issue for the test coverage issue of App Engine code samples. https://github.com/GoogleCloudPlatform/python-docs-samples/issues/7906.

I would prefer merge this PR as an "intermediate milestone", and work with Ace's team on the test coverage of App Engine code samples.

**Comment by jinglundong** [comment]
I've filed https://github.com/GoogleCloudPlatform/python-docs-samples/issues/7906 to track the follow-up work. Is there anything else blocking the approval of this PR?

**Comment by ace-n** [comment]
@jinglundong not that I'm aware of.

'll take a closer look this afternoon. 🙂

File: appengine/standard_python3/bundled-services/deferred/flask/main.py
Comment (ace-n) [inline appengine/standard_python3/bundled-services/deferred/flask/main.py:57]: I assume `/custom/path` is the correct value, and not a placeholder?

File: appengine/standard_python3/bundled-services/mail/django/main.py
Comment (ace-n) [inline appengine/standard_python3/bundled-services/mail/django/main.py:19]: Should these be combined onto one line, or no?

(I also don't know if `flake8` has opinions here.)

_(Apply this comment throughout your PR, if necessary.)_

File: appengine/standard_python3/bundled-services/mail/flask/main.py
Comment (ace-n) [inline appengine/standard_python3/bundled-services/mail/flask/main.py:26]: Should we use a more generic `sender` value here?

**Review by ace-n — APPROVED** [review]
Some minor questions/nits, but LGTM once addressed.

Would it be worth having @engelke take a look as well?

File: appengine/standard_python3/bundled-services/deferred/flask/main.py
Comment (engelke) [reply appengine/standard_python3/bundled-services/deferred/flask/main.py:57]: It is. I could be almost anything, actually. It's the address for an internal POST for deferred handling. The app sets the path and the handler both to /custom/path.

Note that this is not intended to serve external requests at this path.

File: appengine/standard_python3/bundled-services/mail/django/main.py
Comment (engelke) [reply appengine/standard_python3/bundled-services/mail/django/main.py:19]: These are all from separate locations, so I think this is good as is.

File: appengine/standard_python3/bundled-services/mail/flask/main.py
Comment (engelke) [reply appengine/standard_python3/bundled-services/mail/flask/main.py:26]: The django app does not try to send an invalid message, unlike the flask and wsgi versions. All three versions need to be made functionally as close to identical as possible. Note that the sender probably needs to be set up as a valid sending address in an allowlist, so it's hard to see how it can be hard-coded. And the invalid receiving address should be something guaranteed to bounce email.

**Comment by engelke** [comment]
None of the "mail" examples print the body correctly. The wsgi one comes closest, printing the base64 encoded version. This needs to be fixed before merging.

**Review by engelke — CHANGES_REQUESTED** [review]
LGTM once the mail logging and bouncing issues are resolved.

**Pull request closed without merging**
